internal/daemon/api_files.go

```diff
@@ -0,0 +1,595 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// TODO(benhoyt) - should we limit the file size (or total size) for a read-files response,
+//                 or just keep push bytes till the client rejects it?
+// TODO(benhoyt) - should we limit the file size (or total size) for a write-files request,
+//                 or just keep receiving bytes till the disk fills up?
+
+package daemon
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func v1GetFiles(_ *Command, req *http.Request, _ *userState) Response {
+	query := req.URL.Query()
+	action := query.Get("action")
+	switch action {
+	case "read":
+		paths := query["path"]
+		if len(paths) == 0 {
+			return statusBadRequest("must specify one or more paths")
+		}
+		return readFilesResponse{paths: paths}
+	case "list":
+		pattern := query.Get("pattern")
+		if pattern == "" {
+			return statusBadRequest("must specify pattern")
```

Comment:
I guess we're not defaulting to returning everything if pattern is ""?

Reply:
Well, you have to specify the directory, e.g., `pattern=/etc`